src/wcl/doc.h

```diff
@@ -62,10 +73,22 @@ struct doc_state {
   static doc_state from_string(const std::string& str) {
     doc_state state;
 
-    state.character_count = str.size();
+    state.byte_count = str.size();
+
+    utf8proc_int32_t codepoint;
+    utf8proc_ssize_t pos = 0;
+    while (true) {
```

Comment:
I don't think we need to do this but an adapter range for std::string to iterate over code points would be nice. Too much work for right now I think.

Reply:
yeah, that would be a good future improvement. this is a common enough use case